[CS] Add contextual Bool type to switch-case `where` clause

hamishknight opened the pull request:
Seems this was accidentally omitted, which would crash in CSApply for any non-Bool-convertible type.

Resolves #63746

hamishknight commented:
@swift-ci please test

AnthonyLatsis commented on test/Constraints/diagnostics.swift:
```diff
@@ -1531,6 +1531,16 @@ func testUnwrapFixIts(x: Int?) throws {
+func issue63746() {
+  let fn = {
+    switch 0 {
+      case 1 where 0: // expected-error {{type 'Int' cannot be used as a boolean; test for '!= 0' instead}}
```
As an aside, this fix-it is useless for literals.

hamishknight replied:
Heh yeah, filed https://github.com/apple/swift/issues/63753

hamishknight committed:
[CS] NFC: Fix some bad formatting

hamishknight committed:
[CS] NFC: Default the `allowFreeTypeVariables` parameter

All but one client wants to set this to anything
other than `FreeTypeVariableBinding::Disallow`.

hamishknight committed:
[CS] Add contextual Bool type to switch-case `where` clause

Seems this was accidentally omitted, which would
crash in CSApply for any non-Bool-convertible type.

hamishknight commented:
@swift-ci please test
